platforms/android/tangram/src/main/java/com/mapzen/tangram/MapController.java

```diff
@@ -204,19 +204,32 @@ void UIThreadInit(@NonNull final GLViewHolder viewHolder, @Nullable final HttpHa
      * Responsible to dispose internals of MapController during Map teardown.
      * If client code extends MapController and overrides this method, then it must call super.dispose()
      */
-    protected synchronized void dispose() {
-
+    protected void dispose() {
         Log.v(BuildConfig.TAG, "Start MapController dispose");
         nativeMap.shutdown();
         Log.v(BuildConfig.TAG, "Remaining HTTP requests: " + httpRequestHandles.size());
 
         Log.v(BuildConfig.TAG, "MapData instances to remove: " + clientTileSources.size());
         for (MapData mapData : clientTileSources.values()) {
-            mapData.remove();
+            nativeMap.removeClientDataSource(mapData.pointer);
+            mapData.dispose();
         }
         clientTileSources.clear();
         markers.clear();
 
+        // NativeMap dispose makes GL calls to free GPU resources so we _should_ call it on the
+        // render thread, where the GL context is active. However, it is possible for the render
+        // thread associated with GLSurfaceView to be stopped without running all of the queued
+        // events. This would cause the native memory to be leaked.
+        //
+        // To avoid this, we dispose the NativeMap on the UI thread. The GL calls produce an
```

Comment:
I vaguely remember a similar thing when having a texture view setup for the gl context. Maybe update the comment to use generic gl context(y) instead of specific to glsurfaceview.

Reply:
Ahh I see you moved the comment (I added) here. So note to past me to do that. Thanks Matt